### DISCUSSION TITLE
`useState` Array Not Trigger Rerending Using setXXX ([...old]) and useSWR

### DISCUSSION THREAD

**Discussion by EhanDuan** [vercel/next.js#38352] [Help] [answered]
# Overview
Hi, I am building a webapp using `nextjs`, `firesbase`, `chakra UI`, data fetching with `useSWR`.

# Question
I declared 2 arrays like below: 
```ts
  const [focusSkillsIds, setFocusSkillsIds] = useState<string[]>([]);
  const [mainCardsData, setMainCardsData] = useState<mainCardDataType[]>([]);
```
When I updated the 2 arrays, it did not trigger rerendering:
(core code for updating 2 states)
```ts
 useEffect(() => {
    if (data && data.userData && data.userData.focus) {
      data.userData.focus.forEach((sId: string) => {
        if (!focusSkillsIds.includes(sId)) {
          focusSkillsIds.push(sId);
          getMainCardData(sId).then((data) => {
            const currData: mainCardDataType = {
              sId: sId,
              skillName: skillsRoaster.get(sId),
              setId: data.setId,
              setName: data.setName,
              totalCount: data.totalCount,
              solvedCount: data.solvedCount
            };
            console.log(`pushing into mainCardsData @ ${performance.now()}`);   // line 81
            mainCardsData.push(currData);
            console.log(currData);   // line 83
            console.log(mainCardsData); // line 84
          });
        }
      });
    }
    console.log(`trying to update array at ${performance.now()} `); // line 89
    setFocusSkillsIds([...focusSkillsIds]); 
    setMainCardsData([...mainCardsData]);
    // eslint-disable-next-line
  }, [data]);

  if (!data) return <Spinner />;
  if (error) return <Heading>Fail to Load Data</Heading>;
  console.log(`check two arrays at ${performance.now()} `); // line 96
  console.log(focusSkillsIds); 
  console.log(mainCardsData);
  console.log(mainCardsData[0]); 
```

Here are the timeline of the console:
<img width="1757" alt="image" src="https://user-images.githubusercontent.com/75397936/177435557-85e8f65a-abab-4144-b665-a96e3b415f49.png">
<img width="1208" alt="image" src="https://user-images.githubusercontent.com/75397936/177436102-f9b3efe5-76b2-4e6b-b5ea-dbc79ba77875.png">

### At Very First Frame
Both two arrays had the expected data. I think it is from the `useSWR` cache last round. 

**Question 1 : It seems there were data. But the console result of `mainCardsData[0]` was undefined. It did not make sense array had result, but the element in it was undefined.**

### Second Frame Start
Then, the current round started, `useSWR` tried to fetch the new data. At this moment, data from `useSWR` still Empty. Then the `mainCardsData` state array was updated into Empty.

### Second Frame End
Finally, current `useSWR` got the data.  Pushed data into the `mainCardsData` state, but it did not trigger the rerender.

**Question 2: Why here did not trigger rerendering?** 

# Code
```tsx
interface mainCardDataType {
  sId: string | undefined;
  skillName: string | undefined;
  setId: string | undefined;
  setName: string | undefined;
  totalCount: number | undefined;
  solvedCount: number | undefined;
}
const Practice = ({
  sIds,
  skillsData
}: {
  sIds: string[];
  skillsData: any[];
}) => {
  const user = useAuth().user;
  const router = useRouter();
  const [focusSkillsIds, setFocusSkillsIds] = useState<string[]>([]);
  const [mainCardsData, setMainCardsData] = useState<mainCardDataType[]>([]);

  async function getMainCardData(sId: string) {
    const data = await getUnfinishedSetData(user.uid, sId);
    return data;
  }

  const { data, error } = useSWR<any, string>(
    user ? `/api/user/${user.uid}` : null,
    fetcher
  );

  // key -> value :    sId -> name
  const skillsRoaster = useMemo(() => {
    const map = new Map<string, string>();
    sIds.forEach((sId: string, index: number) => {
      map.set(sId, skillsData[index].name);
    });
    return map;
  }, [sIds, skillsData]);

  useEffect(() => {
    if (data && data.userData && data.userData.focus) {
      data.userData.focus.forEach((sId: string) => {
        if (!focusSkillsIds.includes(sId)) {
          focusSkillsIds.push(sId);
          getMainCardData(sId).then((data) => {
            const currData: mainCardDataType = {
              sId: sId,
              skillName: skillsRoaster.get(sId),
              setId: data.setId,
              setName: data.setName,
              totalCount: data.totalCount,
              solvedCount: data.solvedCount
            };
            console.log(`pushing into mainCardsData @ ${performance.now()}`);
            mainCardsData.push(currData);
            console.log(currData);
            console.log(mainCardsData);
          });
        }
      });
    }
    console.log(`trying to update array at ${performance.now()} `);
    setFocusSkillsIds([...focusSkillsIds]);
    setMainCardsData([...mainCardsData]);
    // eslint-disable-next-line
  }, [data]);

  if (!data) return <Spinner />;
  if (error) return <Heading>Fail to Load Data</Heading>;
  console.log(`check two arrays at ${performance.now()} `);
  console.log(focusSkillsIds);
  console.log(mainCardsData);

  if (!mainCardsData || mainCardsData.length == 0) return <Spinner />;

  return (....)

```

**Comment by EhanDuan** [comment]
Just add some additional information

If I change the `useSWR` into a normal `fetcher`, it is the same result.

**Comment by ryanhaticus** [comment]
I'm not entirely sure if this will solve your problem, but consider the following:

```jsx
const currData: mainCardDataType = {
  sId: sId,
  skillName: skillsRoaster.get(sId),
  setId: data.setId,
  setName: data.setName,
  totalCount: data.totalCount,
  solvedCount: data.solvedCount
}
setMainCardsData([currData, ...mainCardsData]);
```

Moreover, it's generally not a good idea to update your state if nothing has truly changed - unnecessary rendering can be devastating for performance, especially in large applications.

**Reply by EhanDuan** [reply]
Yeah, this function indeed worked. I changed a lot including using this your code. Thanks! I am trying to understand what happened under the hood.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The issue here is that you are dealing with async work. It's the core issue anyway, but I see a bunch of problems with the `useEffect`, for example, you disabled the lint rule that is there to help you out!

It might also be a good idea to modify `getMainCardData`, to take in the `user.id` and then move `getMainCardData` outside the React component.

Now to the point. The `useEffect` as you have it can be interpreted as:

```ts
 useEffect(() => {
    data.userData.focus.forEach((sId: string) => {
      getMainCardData(sId).then((data) => {
        const currData: mainCardDataType = {
          sId: sId,
          skillName: skillsRoaster.get(sId),
          setId: data.setId,
          setName: data.setName,
          totalCount: data.totalCount,
          solvedCount: data.solvedCount,
        };
        console.log(`pushing into mainCardsData @ ${performance.now()}`);
        mainCardsData.push(currData);
        console.log(currData);
        console.log(mainCardsData);
      });
    });
    console.log(`trying to update array at ${performance.now()} `);
    setFocusSkillsIds([...focusSkillsIds]);
    setMainCardsData([...mainCardsData]);
  }, [data]);
```

- I've removed the `if` statements to help us see the issue.

As you can see, you start `getMainCardData` calls, which are async work, meaning that the `then` callback executes after the synchronous code, and when the `request` resolves, AKA, the `the` callback is not executed at this point.

Then comes the `` console.log(`trying to update array at ${performance.now()}`); ``, and then you `setMainCardsData` to a shallow copy of its own value, which is not included in the `useEffect` dependencies anyway, so there's a compounded issue here.

The `useEffect` callback ends.

Now the `getMainCardData` callback kicks in, for each `sId`, and they start to mutate the React state, which is an anti-pattern. We are in trouble at this point, this won't render at all! **This is the root of your issue.**


Other things to consider:

- Cancelling the async work when the useEffect dependencies change, AbortController
- Working with accumulators and the callback of `setState` (second argument of `useState`), `setState(prev => [...prev, something])`

### Early termination and scope flattening

This series of nested scopes

```js
function fn(something, foo, bar) {
  if (something) {
    if (foo) {
      if (bar) {
        console.log("hi");
      }
    }
  }
}
```

is often best rewritten as:

```js
function fn(something, foo, bar) {
  if (!something) return;
  if (!foo) return;
  if (!bar) return;

  console.log("hi");
}
```

**Reply by EhanDuan** [reply]
Thanks!
+ Your explanation of the `root Issue` solved my doubts clearly. I also borrowed the idea from @ryanhaticus.  Now I update the states in the `then()` block instead of the previous after it.

+ I also really appreciate your suggestion of **modifying the `getMainCardData`**. It helps with readability and reusability. In addition, `Early termination and scope flattening` is great!

## Doubt on `UseEffect` Deps
Usually, some variables are involved in `useEffect` block, but we do not want to add into `deps`. It seems like the react wants us to add them. It is not complained by complier.

So, I am wondering : should we always follow the rule that   "all variables should in `useEffect` deps" ? Or does it depends on case?
If some variables should not in `useEffect` block, does it mean the logic of the block should be corrected?

**Reply by ryanhaticus** [reply]
Glad this was resolved for you. Happy coding! :)

**Reply by icyJoseph** [reply]
> So, I am wondering : should we always follow the rule that "all variables should in useEffect deps" ? Or does it depends on case?
If some variables should not in useEffect block, does it mean the logic of the block should be corrected?

Always include them. `router.push`/`router.replace` might be a bit of an issue, but otherwise, always include them.

There's patterns to tackle `useEffect`, first of all, always think that the reason to include variables is so that if they change, we re-trigger the effect, but also cancel the ongoing work. `fetch` can be cancelled with `AbortController`, and timers can be cleared, etc.

**Reply by EhanDuan** [reply]
Thanks, it helps a lot!
